File: src/main/java/com/chattriggers/ctjs/launch/mixins/transformers/NoteBlockMixin.java
package com.chattriggers.ctjs.launch.mixins.transformers;

import com.chattriggers.ctjs.minecraft.listeners.CancellableEvent;
import com.chattriggers.ctjs.triggers.TriggerType;
import net.minecraft.block.BlockState;
import net.minecraft.block.NoteBlock;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3f;
import net.minecraft.world.World;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
import org.spongepowered.asm.mixin.injection.callback.LocalCapture;

@Mixin(NoteBlock.class)
public class NoteBlockMixin {
    @Inject(
        method = "onUse",
        at = @At("HEAD"),
        cancellable = true
    )
    public void injectOnUse(
        BlockState state,
        World world,
        BlockPos pos,
        PlayerEntity player,
        Hand hand,
        BlockHitResult hit,
        CallbackInfoReturnable<ActionResult> cir
    ) {
        CancellableEvent event = new CancellableEvent();

        TriggerType.NoteBlockChange.triggerAll(
            new Vec3f((float) pos.getX(), (float) pos.getY(), (float) pos.getZ()),
            // TODO("fabric"): Note name and octave?
            event
        );

        if (event.isCancelled()) {
            cir.setReturnValue(ActionResult.PASS);
            cir.cancel();
        }
    }

    @Inject(
        method = "onSyncedBlockEvent",
        at = @At("HEAD"),
        cancellable = true
    )
    public void injectOnSyncedBlockEvent(
        BlockState state,
        World world,
        BlockPos pos,
        int type,
        int data,
        CallbackInfoReturnable<Boolean> cir
    ) {
        CancellableEvent event = new CancellableEvent();

        TriggerType.NoteBlockPlay.triggerAll(
            new Vec3f((float) pos.getX(), (float) pos.getY(), (float) pos.getZ()),
            // TODO("fabric"): Note name and octave?
            event
        );

        if (event.isCancelled()) {
            cir.setReturnValue(false);
            cir.cancel();
        }
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/launch/mixins/transformers/NoteBlockMixin.kt
package com.chattriggers.ctjs.launch.mixins.transformers

import com.chattriggers.ctjs.triggers.TriggerType
import net.minecraft.block.BlockState
import net.minecraft.block.NoteBlock
import net.minecraft.block.enums.Instrument
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3f
import net.minecraft.world.World
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import org.spongepowered.asm.mixin.injection.callback.LocalCapture

@Mixin(NoteBlock::class)
class NoteBlockMixin {
    @Inject(
        method = ["onUse"],
        at = [At("HEAD")],
        cancellable = true,
        locals = LocalCapture.CAPTURE_FAILHARD,
    )
    fun injectOnUse(
        noteBlock: NoteBlock,
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockHitResult,
        cir: CallbackInfoReturnable<ActionResult>,
    ) {
        TriggerType.NoteBlockChange.triggerAll(
            Vec3f(pos.x.toFloat(), pos.y.toFloat(), pos.z.toFloat()),
            // TODO("fabric"): Note name and octave?
            cir,
        )
    }

    @Inject(
        method = ["onSyncedBlockEvent"],
        at = [At("HEAD")],
        cancellable = true,
        locals = LocalCapture.CAPTURE_FAILHARD,
    )
    fun injectOnSyncedBlockEvent(
        noteBlock: NoteBlock,
        state: BlockState,
        world: World,
        pos: BlockPos,
        type: Int,
        data: Int,
        cir: CallbackInfoReturnable<Boolean>,
    ) {
        // val instrument = state.get(NoteBlock.INSTRUMENT) as Instrument
        // val note = state.get(NoteBlock.NOTE) as Int

        TriggerType.NoteBlockPlay.triggerAll(
            Vec3f(pos.x.toFloat(), pos.y.toFloat(), pos.z.toFloat()),
            // TODO("fabric"): Note name and octave?
            cir,
        )
    }
}
